### ISSUE TITLE
Integration with http://www.blevesearch.com/ for index

### ISSUE THREAD

**Issue by antonikonovalov** [camlistore/camlistore#580]
It's possible?

**Comment by mpl** [comment]
Yes, when we do start thinking about full-text search, this could be one of the possible solutions. (we've seen the talk at fosdem).

**Comment by edrex** [comment]
Bleve also supports numerical facets as well as some geospatial indexing (although that's not documented so not sure how fleshed out it is). Would it be a win to replace some of the current indexing with a bleve-based equivalent?

http://www.blevesearch.com/docs/Index-Mapping/
https://github.com/blevesearch/bleve/tree/master/geo

**Comment by willnorris** [comment]
fwiw, we used bleve to power the search on https://opensource.google.com/projects/.  

It's been a year ago now, so I don't remember all the details, but I know we ran into some issues with weighting fields properly so that matches in titles ranked higher than matches in description, for example.  We also had to do more work than I would have expected to customize some of the query syntax.  All said however, I'd still use it again.

**Comment by gedw99** [comment]
@willnorris  is the bleve code you used on https://opensource.google.com/projects/ open ??

would like to look at how you did it..

**Comment by willnorris** [comment]
> @willnorris is the bleve code you used on opensource.google.com/projects open ??

perhaps ironically, no it's not.  Our end-to-end workflow for populating the project directory is pretty specific to Google infrastructure, and likely wouldn't be something we would release.  There may be bits and pieces of our use of bleve that I could release.  I'll look into it.

**Comment by davex25** [comment]
Has any work been done on this?  I was thinking of working on it, but after a couple hours perusing Perkeep server UI source I still can't figure out how to do something basic like iterate through all images attrs to add to a Bleve index.

**Comment by zenhack** [comment]
I started looking into this myself. One hitch is that bleve has some of its own constraints re: where it wants to keep its storage:

https://github.com/blevesearch/bleve/issues/1137

I don't feel like I understand perkeep's internals well enough to know how much of a problem that inflexibility would be. Do other folks have a sense of this?

**Comment by edrex** [comment]
I think that issue was more discouraging sharing the index storage between different bleve instances.

**Comment by zenhack** [comment]
Hm, the one snag I still see now is that it doesn't look like the perkeep interface satisfies this bit:

>  But, second, bleve must be able to perform multiple read operations from the same stable snapshot. This is where most NoSQL solutions have limitations and are not a good fit.

It doesn't look like perkeep's `KeyValue` interface provides any way of doing multiple reads atomically. I suppose we could just add this to the interface and to each implementation; skimming the source it seems like this should be feasible for all of the existing ones. Thoughts on this approach?

**Comment by mpl** [comment]
if it's doable for all the backends (or at least the most popular ones, i.e. MySQL and levelDB), then I think it's worth a shot. And actually, it does not even have to be doable for all of them. Instead of adding it directly to the KeyValue interface, it could be an additional optional interface that each implementation satisfies or not. Then having Bleve is just conditioned by whether the user chose one of the compatible indexes, which seems fair enough to me.

**Comment by zenhack** [comment]
I looked a little more closely to convince myself, this should indeed be
quite trivial to add to all of them -- except for the kvfile and buffer
backends, all of them are built on transactional data stores, which
makes this straightforward. kvfile and buffer are already using a lock
internally, easy enough to extend that to this use case.

I can take a stab at this soon. From there we can figure out next steps.

**Comment by zenhack** [comment]
I started some work on this here:

https://perkeep-review.googlesource.com/c/perkeep/+/19266

Per the commit message, it's just stubs & a test; I wanted to get feedback on the interface before going ahead and adding the implementations.

**Comment by edrex** [comment]
Checking my understanding:

- Bleve needs read transactions, but not write.
- You can implement read transactions with a https://golang.org/pkg/sync/#RWMutex which won't block on read but will block on write if there are open readers. This is ok as long as users don't expect to be able to write inside a read transaction.

**Comment by edrex** [comment]
* assuming all writes go through the mutex
* for all other existing kv impls the underlying store has transactions

**Comment by zenhack** [comment]
That's a good summary of my own understanding, yes.

**Comment by zenhack** [comment]
Correction: bleve does want write transactions, but the API it expects is already very similar to what sorted.KeyValue currently provides; see NewBatch/ExecuteBatch

https://godoc.org/github.com/blevesearch/bleve/index/store#KVWriter

Note however that it doesn't need read-write transactions -- they're either read-only or write-only.

**Comment by zenhack** [comment]
I sat down and implemented the interfaces for the SQL backend(s). still need to do the others.

**Comment by zenhack** [comment]
Now that the read transaction stuff is merged, I started working on implementing bleve's KVStore on top of the new TransactionalReader interface.

It's mostly implemented, but I hit one snag: the KVBatch interface has a 'Merge' method, which I don't think can be implemented straightforwardly on top of the existing sorted.BatchMutation:

https://godoc.org/github.com/blevesearch/bleve/index/store#KVBatch

It needs to be able to base new values on the existing ones, which means it *does* need to do a read.

Spitballing an extension to the BatchMutation interface, adding a new method:

```go
// Update uses the provided function to modify a key. If the key is not present, the `value` parameter
// will be nil.
func Update(f func(key string, value *string) string)
```

Thoughts?

**Comment by zenhack** [comment]
Hrm, I'm looking at this and I don't see a nice way to implement that *portably* on top of the stuff in sorted. the sql backends start (write) transactions on `BeginBatch()`, so on databases where that means acquiring a lock, if we want to do any reads we have to use the transaction. On the other hand, leveldb has a separate batch write operation, which the current interface maps well to, but doesn't support reads. It does have proper transactions, but there are perf warnings all over it.

Bleve has its own leveldb backend, using the same underlying library as us. What it does for the merges is that it buffers them, and then when doing a commit, first reads the keys that it has merges for, and then adds Set()s to the write batch. It could read a stale value this way, but I guess that's not a problem the way bleve uses it? There's this paragraph in the bleve docs:

> KVWriter is an abstraction for mutating the KVStore KVWriter does **NOT** enforce restrictions of a single writer if the underlying KVStore allows concurrent writes, the KVWriter interface should also do so, it is up to the caller to do this in a way that is safe and makes sense

So maybe we can just buffer the merges, and then after calling CommitBatch, get the values and do the merges, non-atomically?

**Comment by zenhack** [comment]
I went ahead and implemented the non-atomic strategy I described, and moved to next steps of actually hooking this up to bleve, where I hit another wall: The way indices are created seems to only support registering a global driver name and then unmarshalling from a json config, see:

https://godoc.org/github.com/blevesearch/bleve/registry#RegisterKVStore

It seems like not using the local filesystem at all is just a completely unsupported use case; we might be able to beat it into submission, but I'm beginning to think this is not the best approach. My WIP is here:

https://github.com/perkeep/perkeep/blob/af2ecd5e5a3374a8af8b9c35142bc23ed05463a1/pkg/sorted/bleve/kvstore.go

My current thinking is instead we extend the existing Index type in pkg/index to use a bleve index, if available, for full text search. The perkeep config can grow a "bleve" option that specifies the location of bleve's storage. I'm going to start playing around with this approach.

There are other advantages to this beyond just not having to fight with bleve re: where to store the data. For one, it will probably also be faster, both because bleve's default backend doesn't use the pluggable KVStores anyway, and because it means not doing a bunch of conversion between strings and bytes, conversion to hex and soforth to massage sorted.KeyValue into store.KVStore.
